docs/_tutorials/automatic-tensor-parallelism.md

```diff
@@ -0,0 +1,129 @@
+---
+title: "Automatic Tensor Parallelism for HuggingFace Models"
+tags: inference
+---
+
+This tutorial demonstrates the new automatic tensor parallelism feature for inference. Previously, the user needed to provide an injection policy to DeepSpeed to enable tensor parallelism. DeepSpeed now supports automatic tensor parallelism for HuggingFace models by default as long as kernel injection is not true and an injection policy is not provided. This is convenient for when the injection policy of a model is not known and for improving performance of models without kernel injection support.
+
+```python
+# create the model
+import transformers
+import deepspeed
+pipe = pipeline("text2text-generation", model="google/t5-v1_1-small", device=local_rank)
+# Initialize the DeepSpeed-Inference engine
+pipe.model = deepspeed.init_inference(
+    pipe.model,
+    mp_size=world_size,
+    dtype=torch.float
+)
+output = pipe('Input String')
+```
+
+Previously, to run inference with only tensor parallelism for the models that we don't support kernels, you can pass an injection policy that shows the two specific linear layers on a Transformer Encoder/Decoder layer: 1) the attention output GeMM and 2) layer output GeMM. We need these parts of the layer to add the required all-reduce communication between GPUs to merge the partial results across model-parallel ranks. Below, we bring an example that shows how you can use deepspeed-inference with a T5 model:
+
+```python
+# create the model
+import transformers
+from transformers.models.t5.modeling_t5 import T5Block
+import deepspeed
+pipe = pipeline("text2text-generation", model="google/t5-v1_1-small", device=local_rank)
+# Initialize the DeepSpeed-Inference engine
+pipe.model = deepspeed.init_inference(
+    pipe.model,
+    mp_size=world_size,
+    dtype=torch.float,
+    injection_policy={T5Block: ('SelfAttention.o', 'EncDecAttention.o', 'DenseReluDense.wo')}
+)
+output = pipe('Input String')
+```
```

Comment:
Same comment as above code snippet, let's make it runnable if a user were to copy and paste.

Additionally, would it make more sense to have this example first? We show the old way of doing things where the user must specify the `injection_policy` value and then show that we can achieve the same without the user providing that thanks to the new automatic tensor parallelism.

Reply:
I originally had this example first but @awan-10 suggested I reverse it.